File: backend/app/schemas/lesson.py
from datetime import datetime
from typing import Any, Dict, Optional, List

from pydantic import BaseModel, Field

from models.lesson import LessonStatus, LessonType, QuestionType


class QuestionInfo(BaseModel):
    id: int = Field(..., description="Question ID")
    lesson_id: int = Field(..., description="Lesson ID")
    section_id: int = Field(..., description="Section ID")
    type: QuestionType = Field(..., description="Question type")

    content: Optional[Dict[str, Any]] = Field(default=None, description="Question content")
    audio_url: Optional[str] = Field(default=None, description="Audio URL")
    explanation: Optional[str] = Field(default=None, description="Explanation")

    order_index: int = Field(default=0, description="Ordering index")
    created_at: datetime = Field(..., description="Creation timestamp")

    class Config:
        from_attributes = True


class SectionQuestionsResponse(BaseModel):
    section_id: int = Field(..., description="Section ID")
    questions: list[QuestionInfo] = Field(default_factory=list, description="Questions in the section")


class QuestionAnswerSubmitRequest(BaseModel):
    answer: Dict[str, Any] = Field(..., description="Submitted answer payload")


class LearningState(BaseModel):
    energy: int = Field(..., description="Remaining energy")


class QuestionAnswerSubmitResponse(BaseModel):
    question_id: int = Field(..., description="Question ID")
    section_id: int = Field(..., description="Section ID")
    is_correct: bool = Field(..., description="Whether the submitted answer is correct")
    correct_answer: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Correct answer (only returned when submitted answer is wrong)",
    )
    learning_state: LearningState = Field(..., description="Learning state after submitting the answer")


# Lesson Management Schemas
class LessonCreate(BaseModel):
    """Schema for creating a new lesson."""
    topic_id: int = Field(..., description="Topic ID")
    type: LessonType = Field(..., description="Lesson type")
    title: str = Field(..., min_length=1, max_length=150, description="Lesson title")
    description: Optional[str] = Field(default=None, max_length=1000, description="Lesson description")
    audio_url: Optional[str] = Field(default=None, max_length=512, description="Audio URL")
    image_url: Optional[str] = Field(default=None, max_length=512, description="Image URL")
    content: Optional[Dict[str, Any]] = Field(default=None, description="Lesson content as JSON")
    status: LessonStatus = Field(default=LessonStatus.DRAFT, description="Lesson status")
    order_index: int = Field(default=0, ge=0, description="Ordering index")

    class Config:
        extra = "forbid"


class LessonUpdate(BaseModel):
    """Schema for updating a lesson."""
    type: Optional[LessonType] = Field(default=None, description="Lesson type")
    title: Optional[str] = Field(default=None, min_length=1, max_length=150, description="Lesson title")
    description: Optional[str] = Field(default=None, max_length=1000, description="Lesson description")
    audio_url: Optional[str] = Field(default=None, max_length=512, description="Audio URL")
    image_url: Optional[str] = Field(default=None, max_length=512, description="Image URL")
    content: Optional[Dict[str, Any]] = Field(default=None, description="Lesson content as JSON")
    status: Optional[LessonStatus] = Field(default=None, description="Lesson status")
    order_index: Optional[int] = Field(default=None, ge=0, description="Ordering index")

    class Config:
        extra = "forbid"


class LessonResponse(BaseModel):
    """Schema for lesson API response."""
    id: int = Field(..., description="Lesson ID")
    topic_id: int = Field(..., description="Topic ID")
    created_by: int = Field(..., description="Creator user ID")
    type: LessonType = Field(..., description="Lesson type")
    title: str = Field(..., description="Lesson title")
    description: Optional[str] = Field(default=None, description="Lesson description")
    audio_url: Optional[str] = Field(default=None, description="Audio URL")
    image_url: Optional[str] = Field(default=None, description="Image URL")
    content: Optional[Dict[str, Any]] = Field(default=None, description="Lesson content")
    status: LessonStatus = Field(..., description="Lesson status")
    order_index: int = Field(..., description="Ordering index")
    is_deleted: bool = Field(..., description="Whether lesson is soft-deleted")
    created_at: datetime = Field(..., description="Creation timestamp")

    class Config:
        from_attributes = True


class LessonListResponse(BaseModel):
    """Schema for lesson list response."""
    id: int = Field(..., description="Lesson ID")
    topic_id: int = Field(..., description="Topic ID")
    title: str = Field(..., description="Lesson title")
    type: LessonType = Field(..., description="Lesson type")
    status: LessonStatus = Field(..., description="Lesson status")
    order_index: int = Field(..., description="Ordering index")
    created_by: int = Field(..., description="Creator user ID")
    created_at: datetime = Field(..., description="Creation timestamp")

    class Config:
        from_attributes = True

File: frontend/src/types/lesson.ts
type QuestionType = 'MCQ' | 'MULTIPLE_SELECT' | 'FILL_IN_THE_BLANK' | 'MATCHING' | 'TRANSCRIPT' | 'ORDERING' | 'TRUE_FALSE';


export interface SectionQuestionsResponse {
  section_id: number;
  questions: QuestionInfo[];
}

export interface QuestionInfo {
  id: number;
  lesson_id: number;
  section_id: number;
  type: QuestionType;
  content?: Record<string, any> | null;
  audio_url?: string | null;
  explanation?: string | null;
  order_index: number;
  created_at: string;
}

export interface QuestionAnswerSubmitResponse {
  question_id: number;
  section_id: number;
  is_correct: boolean;
  correct_answer?: Record<string, any> | null;
  learning_state: LearningState;
}

export interface QuestionAnswerSubmitRequest {
  answer: Record<string, any>;
}

export interface LearningState {
  energy: number;
}

export type NextSectionResponse = NextSectionAvailable | NextSectionCompleted;

export interface NextSectionAvailable {
  lesson_id: number;
  section: LessonSectionOut;
}

export interface NextSectionCompleted {
  status: "completed";
  message: string;
}

export interface LessonSectionOut {
  id: number;
  lesson_id: number;
  title: string;
  order_index: number;
  content?: Record<string, any> | null;
  created_at: string;
}

export interface CompleteSectionResponse {
  lesson_id: number;
  section_id: number;
  score: number;
  xp: number;
  streak?: number | null;
}

export interface CompleteSectionRequest {
  score: number;
}

export interface LessonContentResponse {
  id: number;
  title: string;
  type: string;
  content?: Record<string, any> | null;
  audio_url?: string | null;
  image_url?: string | null;
}
